### PR TITLE
Resolves #636. 

### LINKED ISSUE #636: Wrapping data with H5DataIO results in different interfaces for accessing the numerical data when round-tripping
## 1) Bug

When wrapping data in an H5DataIO object to support compression, the data interface when round-tripping is different. For example: when dealing with a TimeSeries, the TimeSeries for write has the actual data in TimeSeries.data.data, but the TimeSeries that is read in has the data in TimeSeries.data.

### Steps to Reproduce
```python
import os
import datetime
import numpy as np
from tempfile import gettempdir
from pynwb import NWBFile, NWBHDF5IO, TimeSeries
from pynwb.form.backends.hdf5.h5_utils import H5DataIO

times = np.arange(0, 10, 0.01)
data = np.sin(times)

ts_out = TimeSeries(
    name='some_data',
    source='numpy.sin',
    data=H5DataIO(data, compression=True, compression_opts=2),
    timestamps=times,
    unit='cm/s')

nwbfile_out = NWBFile(
    source='Data source',
    session_description='test',
    identifier='some_uuid',
    session_start_time=datetime.datetime.now(),
    file_create_date=datetime.datetime.now())

nwbfile_out.add_acquisition(ts_out)

filename = os.path.join(gettempdir(), "test.nwb")

with NWBHDF5IO(filename, mode='w') as io:
    io.write(nwbfile_out)

nwbfile_in = NWBHDF5IO(filename, mode='r').read()
ts_in = nwbfile_in.get_acquisition('some_data')
assert(np.allclose(ts_in.data, ts_out.data.data)) # works, but notice the required difference for accessing data
try:
    assert(np.allclose(ts_in.data, ts_out.data)) # fails with TypeError, as ts_out.data is an HDFDataIO object that doesn't support array behavior
finally:
os.remove(filename)
```
### Environment
Please describe your environment according to the following bullet points.
Linux, Python 3.6, pynwb 0.5.1

## Checklist

- [X] Have you ensured the feature or change was not already [reported](https://github.com/NeurodataWithoutBorders/pynwb/issues/) ?
- [X] Have you included a brief and descriptive title?
- [X] Have you included a clear description of the problem you are trying to solve?
- [X] Have you included a minimal code snippet that reproduces the issue you are encountering?

### PR DESCRIPTION
**Pull request by nicain**
Resolves #636 
Delegation for DataChunk and DataIO, allowing testing on DataIO object and element-wise testing of DataChunkIterator

The delegation pattern for DataIO was discussed here:
https://github.com/NeurodataWithoutBorders/pynwb/issues/636#issuecomment-426742988

This lead to an enhanced test:
https://github.com/NeurodataWithoutBorders/pynwb/pull/644/files#r222473586

I additionally implemented a similiar pattern for the DataChunk, because it allows additional testing on the values that are emitted by DataChunkIterator:

https://github.com/NeurodataWithoutBorders/pynwb/pull/644/files#r222477877

## Motivation

Enhance testability.

## How to test the behavior?
See changes to test_base.py and snippet in https://github.com/NeurodataWithoutBorders/pynwb/issues/636

## Checklist

- [x] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [x] Have you ensured the PR description clearly describes problem and the solution?
- [x] Is your contribution compliant with our coding style ? This can be checked running `flake8` from the source directory.
- [x] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [x] Have you included the relevant issue number using `#XXX` notation where `XXX` is the issue number ?

### CODE REVIEW

File: tests/unit/pynwb_tests/test_base.py [hunk 1]
```diff
@@ -59,22 +60,48 @@ def test_nodata(self):
...
     def test_dataio_list_data(self):
-        data = H5DataIO(list(range(100)))
-        ts1 = TimeSeries('test_ts1', 'unit test test_DataIO', data,
+        num_samples = 100
+        data = list(range(num_samples))
+        ts1 = TimeSeries('test_ts1', 'unit test test_DataIO', H5DataIO(data),
                          'grams', starting_time=0.0, rate=0.1)
-        self.assertEqual(ts1.num_samples, 100)
+        self.assertEqual(ts1.num_samples, num_samples)
+        assert data == list(ts1.data)
```
Comment (nicain) [inline tests/unit/pynwb_tests/test_base.py:68]: This is the new test that is possible with DataIO delegation

File: tests/unit/pynwb_tests/test_base.py [hunk 2]
```diff
@@ -59,22 +60,48 @@ def test_nodata(self):
...
+        ts1 = TimeSeries('test_ts1', 'unit test test_DataIO', data,
+                         'grams', starting_time=0.0, rate=0.1)
+        self.assertEqual(ts1.num_samples, -1)
+        for xi, yi in zip(data, generator_factory()):
+            assert np.allclose(xi, yi)
+
+    def test_dci_data_arr(self):
+
+        def generator_factory():
+            return (np.array([i, i+1]) for i in range(100))
```
Comment (nicain) [inline tests/unit/pynwb_tests/test_base.py:97]: DataChunk delegation allows element-wise testing on the data that DataChunkIterator emits:

**Comment by nicain** [comment]
@JFPerkins FYI I just retested your bug report snippet and confirmed it now runs as expected.

**Review by ajtritt — DISMISSED** [review]

**Comment by nicain** [comment]
@ajtritt I just resolved the flake8 complaint on the last merge.  Can you please re-review?

**Review by ajtritt — APPROVED** [review]

**Pull request merged**